compiler/passes/denormalize.cpp

```diff
@@ -669,3 +676,259 @@ inline bool unsafeExprInBetween(Expr* e1, Expr* e2, Expr* exprToMove,
   }
   return false;
 }
+
+/////////////////////////////////////////////////////////////////////////////
+// undoReturnByRef()
+//
+// Ultimately we want to avoid introducing return by ref.
+// For now, simply undo it as a denormalization step.
+
+///////////
+//
+// transformRetTempDef() will modify 'fn' to replace
+//   =(refArg, something)
+//   return(void)
+// with:
+//   return(something)
+
+class DefInfoRR {
+public:
+  FnSymbol*  fn;           // the function that returns by ref
+  ArgSymbol* refArg;       // the function's formal implementing that
+  bool       skipLnFnArgs; // if true, refArg is fn's 3rd-last formal
+};
+
+static bool isGoodRefArg(FnSymbol* fn,              DefInfoRR& info,
+                         bool      skippedLnFnArgs, Expr*      argDef) {
+  ArgSymbol* refArg = toArgSymbol(toDefExpr(argDef)->sym);
+  if (!refArg->hasFlag(FLAG_RETARG))
+    return false; // give up; maybe look harder?
+  INT_ASSERT(!strcmp(refArg->name, "_retArg"));
+  INT_ASSERT(fn->retType = dtVoid);
+
+  info = { fn, refArg, skippedLnFnArgs };
+  return true;
+}
+
+static bool acceptableDef(FnSymbol* fn, DefInfoRR& info) {
+  // isGoodRefArg() fills 'info'
+  if (! isGoodRefArg(fn, info, false, fn->formals.tail)             &&
+      ! (fn->numFormals() >= 3 &&
+         isGoodRefArg(fn, info, true, fn->formals.tail->prev->prev)) )
+    return false;
+
+  if (info.refArg->type->symbol->hasFlag(FLAG_STAR_TUPLE))
+    return false; // codegen requires star tuples to be passed by ref
+
+  for_SymbolSymExprs(refSE, info.refArg) {
+    if (CallExpr* call = toCallExpr(refSE->parentExpr)) {
+      if (call->isPrimitive(PRIM_NO_ALIAS_SET))
+        continue; // any of these are ok
+      if (! call->isPrimitive(PRIM_ASSIGN))
+        return false; // need more work to handle, ex., PRIM_SET_MEMBER
+      // transformRetTempDef() will assert that we have only one PRIM_ASSIGN
+    }
+  }
+
+  return true;
+}
+
+static void transformRetTempDef(DefInfoRR& info) {
+  SymExpr* refUse = nullptr;
+  for_SymbolSymExprs(refSE, info.refArg) {
+    if (CallExpr* call = toCallExpr(refSE->parentExpr))
+      if (call->isPrimitive(PRIM_NO_ALIAS_SET)) {
+        if (call->numActuals() == 1) call->remove();
+        else refSE->remove();
+        continue;
+      }
+    INT_ASSERT(refUse == nullptr); // expect only a single SE
```

Comment:
👍 This is more useful for actually debugging any problems than some sort of counter.